File: packages/ScreenpunkCore/Sources/ScreenpunkCore/Pairing.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#endif

/// Authenticated pairing fixtures. The short code is a SAS from a transcript
/// MAC, not a cipher and not an encryption key.
public enum PairingLimits: Sendable {
    public static let expirySeconds: TimeInterval = 120
    public static let maxFailedAttempts = 5
    public static let codeDigits = 6
    public static let identityByteCount = 32
    public static let sessionNonceByteCount = 16
    public static let sasInfo = "screenpunk-pairing-sas-v1"
}

public enum PairingRole: String, Sendable, Codable, Equatable {
    case device
    case controller
}

public struct PairingIdentity: Sendable, Equatable, Codable {
    public var role: PairingRole
    public var publicKey: [UInt8]

    public init(role: PairingRole, publicKey: [UInt8]) {
        self.role = role
        self.publicKey = publicKey
    }

    public var isWellFormed: Bool {
        publicKey.count == PairingLimits.identityByteCount
    }
}

public struct PairingTranscript: Sendable, Equatable, Codable {
    public var devicePublicKey: [UInt8]
    public var controllerPublicKey: [UInt8]
    public var sessionNonce: [UInt8]

    public init(devicePublicKey: [UInt8], controllerPublicKey: [UInt8], sessionNonce: [UInt8]) {
        self.devicePublicKey = devicePublicKey
        self.controllerPublicKey = controllerPublicKey
        self.sessionNonce = sessionNonce
    }

    public var canonicalBytes: [UInt8] {
        devicePublicKey + [0] + controllerPublicKey + [0] + sessionNonce
    }
}

public enum PairingFailure: String, Error, Sendable, Equatable {
    case expired
    case rateLimited
    case codeMismatch
    case identityChanged
    case secondOwner
    case invalidIdentity
}

public protocol PairingClock: Sendable {
    var now: Date { get }
}

public struct FixedClock: PairingClock, Sendable {
    public var now: Date
    public init(_ now: Date) { self.now = now }
}

#if canImport(CryptoKit)
public enum PairingSAS: Sendable {
    /// HMAC-SHA256(key: sasInfo, data: device || 0x00 || controller || 0x00 || session)
    /// then first 4 bytes as a 6-digit decimal. Standard MAC, not a home-grown cipher.
    public static func matchingCode(for transcript: PairingTranscript) -> String {
        let key = SymmetricKey(data: Data(PairingLimits.sasInfo.utf8))
        let mac = HMAC<SHA256>.authenticationCode(
            for: Data(transcript.canonicalBytes),
            using: key
        )
        let bytes = Array(mac)
        let raw = (UInt32(bytes[0]) << 24)
            | (UInt32(bytes[1]) << 16)
            | (UInt32(bytes[2]) << 8)
            | UInt32(bytes[3])
        return String(format: "%0\(PairingLimits.codeDigits)d", raw % 1_000_000)
    }
}
#endif

public struct PairingSession: Sendable, Equatable {
    public var transcript: PairingTranscript
    public var expectedCode: String
    public var createdAt: Date
    public var failures: Int
    public var confirmed: Bool
    public var candidateOwner: PairingIdentity

    public init(
        transcript: PairingTranscript,
        expectedCode: String,
        createdAt: Date,
        candidateOwner: PairingIdentity,
        failures: Int = 0,
        confirmed: Bool = false
    ) {
        self.transcript = transcript
        self.expectedCode = expectedCode
        self.createdAt = createdAt
        self.candidateOwner = candidateOwner
        self.failures = failures
        self.confirmed = confirmed
    }
}

/// One owning controller identity per device. No credentials are stored here.
public struct DevicePairingState: Sendable, Equatable {
    public var owner: PairingIdentity?
    public var session: PairingSession?

    public init(owner: PairingIdentity? = nil, session: PairingSession? = nil) {
        self.owner = owner
        self.session = session
    }

    public mutating func begin(
        transcript: PairingTranscript,
        expectedCode: String,
        candidateOwner: PairingIdentity,
        clock: PairingClock
    ) throws {
        guard candidateOwner.role == .controller, candidateOwner.isWellFormed else {
            throw PairingFailure.invalidIdentity
        }
        if let owner, owner != candidateOwner {
            throw PairingFailure.secondOwner
        }
        session = PairingSession(
            transcript: transcript,
            expectedCode: expectedCode,
            createdAt: clock.now,
            candidateOwner: candidateOwner
        )
    }

    public mutating func confirm(code: String, presentedOwner: PairingIdentity, clock: PairingClock) throws {
        guard var session else { throw PairingFailure.expired }
        if clock.now.timeIntervalSince(session.createdAt) > PairingLimits.expirySeconds {
            self.session = nil
            throw PairingFailure.expired
        }
        if session.failures >= PairingLimits.maxFailedAttempts {
            throw PairingFailure.rateLimited
        }
        if presentedOwner != session.candidateOwner {
            throw PairingFailure.identityChanged
        }
        if let owner, owner != presentedOwner {
            throw PairingFailure.secondOwner
        }
        if code != session.expectedCode {
            session.failures += 1
            self.session = session
            if session.failures >= PairingLimits.maxFailedAttempts {
                throw PairingFailure.rateLimited
            }
            throw PairingFailure.codeMismatch
        }
        owner = presentedOwner
        session.confirmed = true
        self.session = session
    }
}

File: sdk/src/pairing.ts
import { createHmac } from "node:crypto";

/** Short authentication string. Not a cipher and not an encryption key. */
export const PAIRING_EXPIRY_SECONDS = 120;
export const PAIRING_MAX_FAILURES = 5;
export const PAIRING_CODE_DIGITS = 6;
export const PAIRING_IDENTITY_BYTES = 32;
export const PAIRING_SESSION_BYTES = 16;
export const PAIRING_SAS_INFO = "screenpunk-pairing-sas-v1";

export type PairingRole = "device" | "controller";

export interface PairingIdentity {
  role: PairingRole;
  publicKey: Uint8Array;
}

export interface PairingTranscript {
  devicePublicKey: Uint8Array;
  controllerPublicKey: Uint8Array;
  sessionNonce: Uint8Array;
}

export type PairingFailure =
  | "expired"
  | "rateLimited"
  | "codeMismatch"
  | "identityChanged"
  | "secondOwner"
  | "invalidIdentity"
  | "busy";

export class PairingError extends Error {
  readonly failure: PairingFailure;
  constructor(failure: PairingFailure) {
    super(failure);
    this.failure = failure;
  }
}

export function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) throw new Error("odd hex");
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i += 1) {
    out[i] = Number.parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  }
  return out;
}

export function bytesToHex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("hex");
}

export function repeatingKey(byte: number, length = PAIRING_IDENTITY_BYTES): Uint8Array {
  return new Uint8Array(length).fill(byte);
}

export function canonicalTranscript(transcript: PairingTranscript): Buffer {
  return Buffer.concat([
    Buffer.from(transcript.devicePublicKey),
    Buffer.from([0]),
    Buffer.from(transcript.controllerPublicKey),
    Buffer.from([0]),
    Buffer.from(transcript.sessionNonce)
  ]);
}

/** HMAC-SHA256(key: sasInfo, data: transcript). Standard MAC, not a home-grown cipher. */
export function matchingCode(transcript: PairingTranscript): string {
  const mac = createHmac("sha256", PAIRING_SAS_INFO).update(canonicalTranscript(transcript)).digest();
  const raw = mac.readUInt32BE(0) % 1_000_000;
  return raw.toString().padStart(PAIRING_CODE_DIGITS, "0");
}

export interface DevicePairingState {
  owner: PairingIdentity | null;
  session: {
    transcript: PairingTranscript;
    expectedCode: string;
    createdAtMs: number;
    failures: number;
    confirmed: boolean;
    candidateOwner: PairingIdentity;
  } | null;
}

export function emptyPairingState(): DevicePairingState {
  return { owner: null, session: null };
}

function identitiesEqual(a: PairingIdentity, b: PairingIdentity): boolean {
  return a.role === b.role && bytesToHex(a.publicKey) === bytesToHex(b.publicKey);
}

export function beginPairing(
  state: DevicePairingState,
  transcript: PairingTranscript,
  candidateOwner: PairingIdentity,
  nowMs: number
): DevicePairingState {
  if (candidateOwner.role !== "controller" || candidateOwner.publicKey.length !== PAIRING_IDENTITY_BYTES) {
    throw new PairingError("invalidIdentity");
  }
  if (state.owner && !identitiesEqual(state.owner, candidateOwner)) {
    throw new PairingError("secondOwner");
  }
  // A live session belongs to its candidate until it completes, is cancelled,
  // or expires; a late begin from someone else cannot swap the code on screen.
  if (
    state.session &&
    !state.session.confirmed &&
    nowMs - state.session.createdAtMs <= PAIRING_EXPIRY_SECONDS * 1000 &&
    !identitiesEqual(state.session.candidateOwner, candidateOwner)
  ) {
    throw new PairingError("busy");
  }
  return {
    owner: state.owner,
    session: {
      transcript,
      expectedCode: matchingCode(transcript),
      createdAtMs: nowMs,
      failures: 0,
      confirmed: false,
      candidateOwner
    }
  };
}

export function confirmPairing(
  state: DevicePairingState,
  code: string,
  presentedOwner: PairingIdentity,
  nowMs: number
): DevicePairingState {
  if (!state.session) throw new PairingError("expired");
  if (nowMs - state.session.createdAtMs > PAIRING_EXPIRY_SECONDS * 1000) {
    throw new PairingError("expired");
  }
  if (state.session.failures >= PAIRING_MAX_FAILURES) {
    throw new PairingError("rateLimited");
  }
  if (!identitiesEqual(presentedOwner, state.session.candidateOwner)) {
    throw new PairingError("identityChanged");
  }
  if (state.owner && !identitiesEqual(state.owner, presentedOwner)) {
    throw new PairingError("secondOwner");
  }
  if (code !== state.session.expectedCode) {
    state.session.failures += 1;
    if (state.session.failures >= PAIRING_MAX_FAILURES) throw new PairingError("rateLimited");
    throw new PairingError("codeMismatch");
  }
  return {
    owner: presentedOwner,
    session: { ...state.session, confirmed: true }
  };
}
